File: piazzabot.rb
require 'slack-ruby-client'

BOT_TOKEN = ENV['PIAZZABOT_TOKEN']

Slack.configure do |config|
  config.token = BOT_TOKEN
end

$cl = Slack::RealTime::Client.new

def start_with_retry!
  loop do
    begin
      $cl.start!
      break
    rescue StandardError
      sleep 15
      puts "Retrying at #{Time.now}"
    end
  end
end

$cl.on :hello do
  puts "Running"
  $botid = $cl.self.id
  $botname = $cl.self.name
end

def invalid?(data)
  (data.nil? || data.text.nil? || data.user.nil? || data.channel.nil?)
end

def msg(channel, text)
  $cl.web_client.chat_postMessage channel: channel, text: text, as_user: false, username: "campuswirebot"
end

$cl.on :message do |data|
  begin
    next if invalid?(data)
    
    match = /(?:^|\W)#(\d{1,4})\b/.match(data.text)
    if match
      msg(data.channel, "https://campuswire.com/c/G8FE6D09F/feed/#{match[1]}")
    end
  rescue StandardError => e
    exception = ([e.inspect] + e.backtrace.select { |line| !line.include? ".rvm" }).join("\n")
    msg(data.channel, "<@U5G7Z8NM9> #{exception}")
  end
end

$cl.on :close do
  puts "Closing"
end

$cl.on :closed do
  puts "Closed"
  start_with_retry!
end

start_with_retry!

File: piazzabot.py
import slackclient
import re
import websocket
import os
import time
import traceback

BOT_TOKEN = os.environ['PIAZZABOT_TOKEN']
DIAGNOSTIC_CHANNEL = "#_test"  # where to post update and crash recovery notices


class Piazzabot:
  def __init__(self):
    self.cli = slackclient.SlackClient(BOT_TOKEN)
    assert self.cli.rtm_connect()
    auth = self.cli.api_call("auth.test")
    self.bot_id = auth['user_id']
    self.bot_name = auth['user']

  def post(self, msg, channel):
    return self.cli.api_call("chat.postMessage",
      channel=channel,
      text=msg,
      as_user=False,
      username="campuswirebot")
  
  def run(self, recovered_traceback):
    print('Running')
    if recovered_traceback:
      self.post(recovered_traceback, DIAGNOSTIC_CHANNEL)
    while True:
      for output in self.cli.rtm_read():
        if output['type'] == 'message':
          m = re.search(r"(?:^|\W)#(\d{1,4})\b", output.get('text', ''))
          if m:
            self.post(f"https://campuswire.com/c/G8FE6D09F/feed/{m.group(1)}", output['channel'])
      time.sleep(1)


if __name__ == '__main__':
  # on crash, wait a little then reconnect the bot
  recovered_traceback = None
  while True:
    try:
      Piazzabot().run(recovered_traceback)
    except (slackclient.server.SlackConnectionError, websocket._exceptions.WebSocketException):
      # happens at random, unavoidable, don't bother saving the stack trace
      time.sleep(1)
    except Exception:
      # 'actual' problem; save and post it upon reconnecting
      traceback.print_exc()
      recovered_traceback = traceback.format_exc()
      time.sleep(1)

